### DISCUSSION TITLE
Throttle/Debounce @bind elements

### DISCUSSION THREAD

**Discussion by bvdmitri** [fonsp/Pluto.jl#1114] [Pluto development]
Discussion from Slack#pluto channel.

Basically the idea is to have a proper API to debounce and/or throttle `@bind` elements. Sometimes it is inefficient to update the state of the notebook on the change rate of e.g. `Slider` from `PlutoUI.jl`. It might be a good feature to have a proper debouncing/throttling, especially because there is no need to update anything faster than `window.requestAnimationFrame`.

Based on @fonsp comment in Zulip I made a gist for temporary solution for throttled slider to use in Pluto notebook:
https://gist.github.com/bvdmitri/5523d1c7ffa13b5a54e0235dd037c4b0 It can be also easily modified to use `debounce`.

**Comment by fonsp** [comment]
Great! Note that Pluto already throttles all inputs: if cells are running, bind events are delayed and previously queued bind events are discarded. Try this:
```julia
@bind x Slider(1:1000)
```
```julia
sleep(1); x
```

But further throttling and debouncing is still useful for more advanced UI.

**Reply by bvdmitri** [reply]
I see, I didn't know about that, it is cool 😎. 

But the main idea of throttling/debouncing is not to discard previous events, but to wait until some condition is satisfied. It actually makes a significant difference in some scenarios, because some code may still consume CPU resources until it is discarded.  Example with sleep does not lag of course because CPU is idle but when we run actual heavy code the whole notebook still lags with slider.

If Pluto already has this feature it would be cool to expose it as a proper API so libraries and UI tools might have a direct access to it and have a control over it.

**Reply by mbauman** [reply]
Is there something I need to do with a custom JS input to get this behavior?  I have a Pluto notebook with a Canvas element you gave me a while back that dispatches a `CustomEvent("input")`.

<details>
<summary>Cell code that does the `@bind`</summary>

```html
@bind img html"""
<canvas width="200" height="200" style="margin: auto; display: block;"></canvas>
<button id="clearcanvas" style="font-size:3ex; margin: auto; display: block">Clear</button>
<script>
const canvas = this.document.querySelector("canvas")
const button = this.document.querySelector("#clearcanvas")
const ctx = canvas.getContext("2d")
function send_image(){
	// 🐸 We send the value back to Julia 🐸 //
	canvas.value = {
		width: 200,
		height: 200,
		data: ctx.getImageData(0,0,200,200).data,
	}
	canvas.dispatchEvent(new CustomEvent("input"))
}
var prev_pos = [0, 0]
function clear(){
	ctx.fillStyle = '#ffecec'
	ctx.fillRect(0, 0, 200, 200)
	send_image()
}
clear()
function onmove(e){
	const new_pos = [e.layerX, e.layerY]
	ctx.lineTo(...new_pos)
	ctx.stroke()
	prev_pos = new_pos
	send_image()
}
canvas.onmousedown = e => {
	prev_pos = [e.layerX, e.layerY]
	ctx.strokeStyle = '#003d6d'
	ctx.lineWidth = 12
	ctx.beginPath()
	ctx.moveTo(...prev_pos)
	canvas.onmousemove = onmove
}
button.onclick = clear
canvas.onmouseup = e => {
	canvas.onmousemove = null
}
</script>
"""
```
</details>

In Pluto v0.12.x, this discarded queued events as you describe, but after upgrading to v0.14 all queued events get processed.  Do I need to change something here?

**Comment by fonsp** [comment]
Here is the draft PR: https://github.com/fonsp/PlutoUI.jl/pull/105

**Reply by fonsp** [reply]
@mbauman
